File: deno/src/cmd/clean.ts
import { dockerCapture } from "../lib/docker.ts";
import { canonicalize } from "../lib/paths.ts";

// Remove every overlay volume tied to a project. Volumes carry the
// maudebox.project label as the canonical link (more durable than the
// volume's name suffix scheme). Doesn't validate that the project still
// exists — cleaning orphan volumes for a since-deleted worktree by
// passing the original path is a legitimate use.
export async function cmdClean(args: string[]): Promise<number> {
    if (args[0] === "-h" || args[0] === "--help") {
        const { USAGE } = await import("../usage.ts");
        console.log(USAGE);
        return 0;
    }
    const projectDir = await canonicalize(args[0] ?? ".");
    const out = await dockerCapture([
        "volume",
        "ls",
        "--filter",
        `label=maudebox.project=${projectDir}`,
        "--format",
        "{{.Name}}",
    ]);
    const volumes = out.split("\n").filter((v) => v.length > 0);
    if (volumes.length === 0) {
        console.log(`No overlay volumes found for: ${projectDir}`);
        return 0;
    }
    for (const v of volumes) {
        await dockerCapture(["volume", "rm", v]);
        console.log(`Removed volume: ${v}`);
    }
    return 0;
}

File: src/cmd/clean.rs
use crate::docker;
use crate::paths::canonicalize;
use anyhow::Result;

// Remove every overlay volume tied to a project. Volumes carry the
// maudebox.project label as the canonical link. Doesn't validate that the
// project still exists — cleaning orphan volumes for a since-deleted
// worktree by passing the original path is a legitimate use.
pub fn run(project_dir: &str) -> Result<i32> {
    let project_dir = canonicalize(project_dir);

    let label = format!("label=maudebox.project={}", project_dir.display());
    let out = docker::capture(&["volume", "ls", "--filter", &label, "--format", "{{.Name}}"])?;
    let volumes: Vec<&str> = out.lines().filter(|l| !l.is_empty()).collect();
    if volumes.is_empty() {
        println!("No overlay volumes found for: {}", project_dir.display());
        return Ok(0);
    }
    for v in volumes {
        docker::capture(&["volume", "rm", v])?;
        println!("Removed volume: {v}");
    }
    Ok(0)
}
